### PR TITLE
E2E test suite manages its preprod node/wallet

### PR DESCRIPTION
**Pull request by Unisay**
- [x] Run tests against different networks depending on the CLI flags.
- [x] E2E test suite manages a preprod node instance.
- [x] Preserve node/wallet state in a local folder to avoid syncing from scratch.
- [x] E2E test suite manages a wallet instance.

### Issue Number

ADP-3137

### CODE REVIEW

**Comment by Unisay** [comment]
```
Tests:

Wallet Backend API
Writing node logs to /home/projects/cardano-wallet/lib/wallet-e2e/test-state/node/node.log
Writing wallet logs to /home/projects/cardano-wallet/lib/wallet-e2e/test-state/wallet/wallet.log
Waiting for the node socket ...
Waiting for the node socket ...
Waiting for the node socket ...
Waiting for the node socket ...
Node sync progress: 100.0%
Node status as reported by Wallet: NodeIsSynced
  ✓ Created wallet is listed                                             730.90 ms
  ✓ Created wallet can be retrieved by id                                207.56 ms
  ✓ Created wallet has zero ADA balance                                  232.59 ms
Effect interpreters
  Timeouts
    ✓ no timeout                                                         901.34 ms
    ✓ timeout in 10 milliseconds                                        1000.68 ms
 
 
  Passed:                       5
  Failed:                       0
  Sum of test runtimes:         3.07 seconds
  Test suite took:              9.65 seconds
  ```

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Query.hs [hunk 1]
```diff
@@ -117,11 +121,12 @@ runQuery
        , FxAssert :> es
        , FxTrace :> es
        )
-    => Eff (FxQuery : es) a
+    => NetworkConfig
+    -> Eff (FxQuery : es) a
     -> Eff es a
-runQuery = interpret \_ -> \case
+runQuery networkConfig = interpret \_ -> \case
```
Comment (paolino) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Query.hs:127]: should we start moving each case in a separate module or at least top level ?

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Node/Cli.hs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    , era :: Text
+    , hash :: Text
+    , slot :: Natural
+    , slotInEpoch :: Natural
+    , slotsToEpochEnd :: Natural
+    , syncProgress :: Double
+    }
+    deriving stock (Eq, Show, Generic)
+
+instance FromJSON NodeTip where
```
Comment (paolino) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Node/Cli.hs]: Can we avoid the instance and use it as a  function ?

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Preprod.hs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+import Cardano.Wallet.Spec.Data.Network.NodeStatus
+    ( NodeStatus (..) )
+import Cardano.Wallet.Spec.Network.Config
+    ( NetworkConfig (..) )
+import Cardano.Wallet.Spec.Network.Node
+    ( NodeProcessConfig (..) )
+import Cardano.Wallet.Spec.Network.Wallet
+    ( WalletProcessConfig (..) )
+import Control.Monad.Trans.Resource
```
Comment (paolino) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Preprod.hs:20]: good!

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Preprod.hs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                , walletByronGenesis =
+                    preprodDir </> [relfile|byron-genesis.json|]
+                }
+    (_walletReleaseKey, (_walletInstance, walletApi)) <-
+        allocate (Wallet.start walletProcessConfig) (Wallet.stop . fst)
+
+    -- Wait for the node to sync with the network
+    let policy = capDelay 10_000_000 (fibonacciBackoff 1_000_000)
+
+    _ <- retrying policy (\_rs -> pure . (< 99.99)) \_ -> liftIO do
```
Comment (paolino) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Preprod.hs]: should we factor out 

```hs
waiting
    :: (a -> Bool)
    -> (RetryStatus -> ResourceT IO a)
    -> ResourceT IO a
waiting f = retrying policy (\_ -> pure . f)
  where
    policy = capDelay 10_000_000 (fibonacciBackoff 1_000_000)
```

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Wallet/Cli.hs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    , networkTip :: NetworkTip
+    , nextEpoch :: NextEpoch
+    , nodeEra :: Text
+    , nodeTip :: NodeTip
+    , nodeStatus :: NodeStatus
+    , walletMode :: Text
+    }
+    deriving stock (Eq, Show)
+
+instance FromJSON NetworkInformation where
```
Comment (paolino) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Wallet/Cli.hs]: Same suggestion, avoid instances ?

**Review by paolino — APPROVED** [review]

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Node/Cli.hs (hunk 2 above)
Comment (Unisay) [reply lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Node/Cli.hs]: Done

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Wallet/Cli.hs (hunk 5 above)
Comment (Unisay) [reply lib/wallet-e2e/src/Cardano/Wallet/Spec/Network/Wallet/Cli.hs]: Done

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Query.hs (hunk 1 above)
Comment (Unisay) [reply lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Query.hs:127]: Soon :)

**Pull request merged**